Error message for missing return value with Result type is unclear

The error message that I get looks like this:

```
src/main.rs:18:9: 18:11 error: unable to infer enough type information about `_`; type annotations required [E0282]
src/main.rs: 18         Ok(Response::with( (status::Ok, payload) ));
error: aborting due to previous error
```

The highlight is on the leading `Ok`, which is misleading. The actual fix was to either precede the statement with `return`, or to drop the trailing `;`. It would be great if the compiler were to suggest this as well, as the current error message alone is cryptic.

Triage: here's today's error, no real change:

```text
error[E0282]: unable to infer enough type information about `_`
 --> <anon>:2:5
  |
2 |     Ok(1);
  |     ^^ cannot infer type for `_`
  |
  = note: type annotations or generic parameter binding required

```

Going to go ahead and mark this as E-needstest since it feels easy to regress on and medium-level bad if we do.

In particular, I think adding to the existing UI test `src/test/ui/coercion-missing-tail-expected-type.rs` is the right thing to do. The example code given [here](https://github.com/rust-lang/rust/issues/24889#issuecomment-96980747) should be added to the existing UI test, the test run, and then the stderr file updated by cd-ing into `src/test/ui` and running the `update-all-references.sh` script. Given these instructions, marking as E-easy and E-mentor.

```
error[E0308]: mismatched types
 --> test.rs:1:29
  |
1 |   fn foo() -> Result<u8, u64> {
  |  _____________________________^
2 | |     Ok(1);
3 | | }
  | |_^ expected enum `std::result::Result`, found ()
  |
  = note: expected type `std::result::Result<u8, u64>`
             found type `()`
help: consider removing this semicolon:
 --> test.rs:2:10
  |
2 |     Ok(1);
  |          ^

error: aborting due to previous error(s)
```

--- a/src/test/ui/coercion-missing-tail-expected-type.rs
+++ b/src/test/ui/coercion-missing-tail-expected-type.rs
@@ -14,6 +14,10 @@ fn plus_one(x: i32) -> i32 {
     x + 1;
 }
 
+fn foo() -> Result<u8, u64> {
+    Ok(1);
+}
+
 fn main() {
     let x = plus_one(5);
     println!("X = {}", x);

--- a/src/test/ui/coercion-missing-tail-expected-type.stderr
+++ b/src/test/ui/coercion-missing-tail-expected-type.stderr
@@ -15,5 +15,22 @@ help: consider removing this semicolon:
 14 |     x + 1;
    |          ^
 
+error[E0308]: mismatched types
+  --> $DIR/coercion-missing-tail-expected-type.rs:17:29
+   |
+17 |   fn foo() -> Result<u8, u64> {
+   |  _____________________________^
+18 | |     Ok(1);
+19 | | }
+   | |_^ expected enum `std::result::Result`, found ()
+   |
+   = note: expected type `std::result::Result<u8, u64>`
+              found type `()`
+help: consider removing this semicolon:
+  --> $DIR/coercion-missing-tail-expected-type.rs:18:10
+   |
+18 |     Ok(1);
+   |          ^
+
 error: aborting due to previous error(s)
 

